File: browser-automation-engine/refactor-charge-job.py
"""
Script untuk refactoring charge job section di _attendance_logic.json
Mengganti parseChargeJob + PART 1-5 dengan include ke _charge_job_full.json
"""

import json
import sys
from pathlib import Path

# Fix Windows encoding
if sys.stdout.encoding != 'utf-8':
    import io
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
if sys.stderr.encoding != 'utf-8':
    import io
    sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8')

template_path = Path(__file__).parent / 'templates' / '_attendance_logic.json'
backup_path = Path(__file__).parent / 'templates' / '_attendance_logic.json.backup'
output_path = Path(__file__).parent / 'templates' / '_attendance_logic_v3.json'

print('=' * 60)
print('  REFACTOR CHARGE JOB SECTION')
print('=' * 60)
print()

# Read template
with open(template_path, 'r', encoding='utf-8') as f:
    content = f.read()
    template = json.loads(content)

def find_all_parse_charge_job(obj, path=""):
    """Find all parseChargeJob actions in the JSON structure"""
    results = []

    if isinstance(obj, dict):
        if obj.get('action') == 'parseChargeJob':
            results.append((path, obj))
        for key, value in obj.items():
            results.extend(find_all_parse_charge_job(value, f"{path}.{key}" if path else key))
    elif isinstance(obj, list):
        for i, item in enumerate(obj):
            results.extend(find_all_parse_charge_job(item, f"{path}[{i}]"))

    return results

print("Searching for parseChargeJob actions...")
found = find_all_parse_charge_job(template)

print(f"\nFound {len(found)} parseChargeJob actions:")
for path, obj in found:
    print(f"  - {path}")

# Now let's replace them
def replace_charge_job_recursive(obj):
    """Recursively replace charge job sections"""
    if isinstance(obj, dict):
        # Check if this is a parseChargeJob action
        if obj.get('action') == 'parseChargeJob':
            print(f"\nFound parseChargeJob, replacing...")
            # Return the include action instead
            return {
                "action": "include",
                "params": {
                    "template": "_charge_job_full"
                },
                "comment": "═══ CHARGE JOB INPUT (extracted to _charge_job_full.json) ═══"
            }

        # Process all values in the dict
        return {k: replace_charge_job_recursive(v) for k, v in obj.items()}

    elif isinstance(obj, list):
        # Check if this list contains parseChargeJob
        new_list = []
        skip_until_marker = False

        for item in obj:
            if skip_until_marker:
                # Check if we should stop skipping
                if isinstance(item, dict):
                    action = item.get('action', '')
                    message = item.get('params', {}).get('message', '')
                    comment = item.get('comment', '')

                    # Stop skipping markers
                    if (action == 'log' and 'Skipping Hour Input' in message) or \
                       (action == 'log' and 'Skipping Enter key' in message) or \
                       (action == 'typeInput' and 'txtHours' in item.get('params', {}).get('selector', '')) or \
                       (action == 'if' and 'VALIDATION' in comment):
                        skip_until_marker = False
                        new_list.append(item)
                    else:
                        continue  # Skip this item
                else:
                    new_list.append(replace_charge_job_recursive(item))
            else:
                if isinstance(item, dict) and item.get('action') == 'parseChargeJob':
                    # Start skipping mode - replace with include
                    new_list.append({
                        "action": "include",
                        "params": {
                            "template": "_charge_job_full"
                        },
                        "comment": "═══ CHARGE JOB INPUT (extracted) ═══"
                    })
                    skip_until_marker = True
                else:
                    new_list.append(replace_charge_job_recursive(item))

        return new_list

    return obj

print("\nProcessing template...")
result = replace_charge_job_recursive(template)

print("\n[OK] Refactoring complete!")

# Create backup if not exists
if not backup_path.exists():
    import shutil
    shutil.copy(template_path, backup_path)
    print(f"[OK] Backup created: {backup_path.name}")

# Write output
with open(output_path, 'w', encoding='utf-8') as f:
    json.dump(result, f, indent=2, ensure_ascii=False)

print(f"[OK] Saved to: {output_path.name}")
print()
print('=' * 60)
print('  NEXT STEPS:')
print('  1. Check _attendance_logic_v3.json')
print('  2. Verify charge job sections are replaced')
print('  3. Test with the automation')
print('=' * 60)

File: browser-automation-engine/refactor-charge-job.js
/**
 * Script untuk refactoring charge job section di _attendance_logic_v2.json
 * Mengganti charge job steps dengan include ke _charge_job_full.json
 */

const fs = require('fs');
const path = require('path');

const templatePath = path.join(__dirname, 'templates', '_attendance_logic_v2.json');
const outputPath = path.join(__dirname, 'templates', '_attendance_logic_v2.json');

console.log('═══════════════════════════════════════════════════');
console.log('  REFACTOR CHARGE JOB SECTION');
console.log('═══════════════════════════════════════════════════\n');

// Read template
const template = JSON.parse(fs.readFileSync(templatePath, 'utf8'));

// Function to replace charge job steps with include
function replaceChargeJobSteps(steps) {
    let replacedCount = 0;

    function processStep(step) {
        // Check if this is a parseChargeJob action
        if (step.action === 'parseChargeJob') {
            // Found the start of charge job section
            return { action: 'include_marker', replace: true };
        }

        // Check if this is one of the charge job input steps
        if (step.comment && step.comment.includes('PART 1: Task Code')) {
            return { action: 'include_marker', replace: true };
        }
        if (step.comment && step.comment.includes('PART 2: Station Code')) {
            return { action: 'include_marker', replace: true };
        }
        if (step.comment && step.comment.includes('PART 3: Machine Code')) {
            return { action: 'include_marker', replace: true };
        }
        if (step.comment && step.comment.includes('PART 4: Expense Code')) {
            return { action: 'include_marker', replace: true };
        }
        if (step.comment && step.comment.includes('PART 5: Additional Code')) {
            return { action: 'include_marker', replace: true };
        }

        // Process nested steps
        if (step.thenSteps) {
            step.thenSteps = processSteps(step.thenSteps);
        }
        if (step.elseSteps) {
            step.elseSteps = processSteps(step.elseSteps);
        }
        if (step.params && step.params.steps) {
            step.params.steps = processSteps(step.params.steps);
        }

        return step;
    }

    function processSteps(steps) {
        const newSteps = [];
        let i = 0;

        while (i < steps.length) {
            const step = steps[i];

            // Check if this step is parseChargeJob or has charge job comment
            if (step.action === 'parseChargeJob' ||
                (step.comment && step.comment.includes('PART ') && step.comment.includes('Code'))) {

                // Found charge job section - replace with include
                newSteps.push({
                    "action": "include",
                    "params": {
                        "template": "_charge_job_full"
                    },
                    "comment": "═══ CHARGE JOB INPUT (extracted to _charge_job_full.json) ═══"
                });

                // Skip all charge job related steps
                let skipped = 0;
                while (i + skipped < steps.length) {
                    const nextStep = steps[i + skipped];
                    if (!nextStep) break;

                    // Stop skipping when we hit a non-charge-job step
                    const isChargeJobStep =
                        nextStep.action === 'parseChargeJob' ||
                        nextStep.action === 'retryInputWithValidation' ||
                        nextStep.action === 'if' && nextStep.comment?.includes('PART') ||
                        nextStep.action === 'wait' && nextStep.comment?.includes('PART') ||
                        (nextStep.comment && (
                            nextStep.comment.includes('PART 1:') ||
                            nextStep.comment.includes('PART 2:') ||
                            nextStep.comment.includes('PART 3:') ||
                            nextStep.comment.includes('PART 4:') ||
                            nextStep.comment.includes('PART 5:')
                        )) ||
                        (nextStep.action === 'wait' && i + skipped + 1 < steps.length &&
                         steps[i + skipped + 1].comment?.includes('PART'));

                    if (!isChargeJobStep) break;

                    skipped++;
                }

                i += skipped;
                replacedCount++;
            } else {
                // Process nested steps
                const processedStep = processStep(step);
                newSteps.push(processedStep);
                i++;
            }
        }

        return newSteps;
    }

    return { steps: processSteps(steps), replacedCount };
}

// Process the template
console.log('Processing template...');
const result = replaceChargeJobSteps(template.steps);
template.steps = result.steps;

console.log(`✅ Replaced ${result.replacedCount} charge job sections`);

// Write back
fs.writeFileSync(outputPath, JSON.stringify(template, null, 2));
console.log(`✅ Saved to: ${path.basename(outputPath)}`);

console.log('\n═══════════════════════════════════════════════════');
console.log('  NEXT: Check _attendance_logic_v2.json');
console.log('  Run: node index.js attendance-input-loop-v2');
console.log('═══════════════════════════════════════════════════\n');
